add support for using set! on dsl-bound variables

quasarbright opened the pull request:
Previously, using set! on a bound id would error out as if you tried to set!
a macro defined by define-syntax, or it would say that compile-reference
expected an identifier and got a set! expression. Now, it works.

Also added `doc/` to `.gitignore`.

Tests are in `private/test/set-bang.rkt`

michaelballantyne commented:
It seems like it might be unnecessarily indirect to compile first with the reference compiler to a name bound as a set! transformer, and then to the ultimate compilation via the expansion of that transformer.

What if we made it so that you could pass a `set!-transformer` value as a reference compiler in `resume-host-expansion`?  For your test case, something like...

```
    [(mylet v:id body:expr)
     (resume-host-expansion #'body #:reference-compilers ([var compile-var/set!]))]
```

```
(define (compile-var/set! id)
  (make-set!-transformer
     (syntax-parser
        [(set! x:id val:expr) #'(list 'set! val)]
        [x:id #''ref])))
```

quasarbright commented:
Fixed a bug where applying dsl-bound variables as functions would fail while invoking the set! transformer. Added a case to the set!-transformer's parser to handle this.

I tried using `make-variable-like-transformer`, but that didn't work because for the set! case, you only get access to the syntax for the new value, not the identifier, which needs to be compiled. Instead, I added a third case for applications and did the same thing that `make-variable-like-transformer` does.

Added tests to `private/test/set-bang.rkt`

quasarbright commented:
> It seems like it might be unnecessarily indirect to compile first with the reference compiler to a name bound as a set! transformer, and then to the ultimate compilation via the expansion of that transformer.
> 
> What if we made it so that you could pass a `set!-transformer` value as a reference compiler in `resume-host-expansion`? For your test case, something like...
> 
> ```
>     [(mylet v:id body:expr)
>      (resume-host-expansion #'body #:reference-compilers ([var compile-var/set!]))]
> ```
> 
> ```
> (define (compile-var/set! id)
>   (make-set!-transformer
>      (syntax-parser
>         [(set! x:id val:expr) #'(list 'set! val)]
>         [x:id #''ref])))
> ```

That sounds like a good idea. It would give the user control over whether these variables can be mutable. A few thoughts:
* Would this be a built-in? Or would it be something that the user has to provide?
* Where would the user specify whether variables are set!-able? Just in the compiler? Or in the binding spec too?
* Would they have to implement this custom reference compiler? Or will this be a library-provided alternative to compile-reference? For the latter, either the binding class or compile-binder! would have to cause a set!-transformer to be put in the expander environment, right?
* Which should the default behavior be? Mutable, or immutable? 
* If the user tries to call set! on an immutable variable, they'd get an error like "cannot mutate a syntax id". If we make a set!-transformer for immutable variables, we can put a more informative error in the set! case.

quasarbright commented:
In the future, we want to allow something like
```racket
(struct my-compile-time-var-rep #:prop prop:match-expander ...)
(define (compile-expr ...) ... (resume-host-expansion ... #:reference-compilers ([var my-compile-time-var-rep])) ... )
```
That's not easily doable right now. For now, we'll support that sort of thing with `prop:set!-transformer` checking if `compile` is a `set!-transformer` and using that instead of the `syntax-parse` if it is. I'll add this change in a new commit on this PR

quasarbright commented:
In a future PR, I can also provide something like `compile-mutable-reference` and `compile-immutable-reference` which has default set! behaviors of propagating set! and raising a clear syntax error respectively.

quasarbright commented:
set!-transformers can now be passed as reference compilers. I had to expand the contract for resume-host-expansion-rt a bit. You might just want to get rid of that when you eventually support arbitrary structs with transformer props.

I implemented `compile-mutable-reference` in my test file, but I didn't change compile-reference or add any new reference compilers to the library. I remember we were talking  about adding compile-immutable-reference and compile-mutable-reference, and I think that's a good idea. But I'll leave that to another PR/commit

michaelballantyne commented:
Thanks! I made some small tweaks in https://github.com/michaelballantyne/bindingspec/commit/f8694e72339778a1470b3b087827ec6116b7a9db and https://github.com/michaelballantyne/bindingspec/commit/e35da0ee0f0c530c38f8e1d20e0a4bd3f7f85296 and merged.

Most substantially, I put the variable-like-transformer handling of application-like uses into new `immutable-reference-compiler` and `mutable-reference-compiler` transformers rather than building into the #:reference-compiler behavior. The previous design was inconsistent: when given a plain procedure this case was handled automatically, but when given a set!-transformer the transformer had to handle it.
